parsing/parser.mly

```diff
@@ -3392,44 +3392,75 @@ tuple_type:
    - applications of type constructors:   int, int list, int option list
    - variant types:                       [`A]
  *)
+
+
+(*
+  Delimited types:
+    - parenthesised type          (type)
+    - first-class module types    (module S)
+    - object types                < x: t; ... >
+    - variant types               [ `A ]
+    - extension                   [%foo ...]
+*)
+delimited_type_without_object_type:
+  | LPAREN type_ = core_type RPAREN
+      { type_ }
+  | LPAREN MODULE attrs = ext_attributes package_type = package_type RPAREN
+      { wrap_typ_attrs ~loc:$sloc (reloc_typ ~loc:$sloc package_type) attrs }
+  | mktyp(
+      LBRACKET field = tag_field RBRACKET
+        { Ptyp_variant([ field ], Closed, None) }
+    | LBRACKET BAR fields = row_field_list RBRACKET
+        { Ptyp_variant(fields, Closed, None) }
+    | LBRACKET field = row_field BAR fields = row_field_list RBRACKET
+        { Ptyp_variant(field :: fields, Closed, None) }
+    | LBRACKETGREATER BAR? fields = row_field_list RBRACKET
+        { Ptyp_variant(fields, Open, None) }
+    | LBRACKETGREATER RBRACKET
+        { Ptyp_variant([], Open, None) }
+    | LBRACKETLESS BAR? fields = row_field_list RBRACKET
+        { Ptyp_variant(fields, Closed, Some []) }
+    | LBRACKETLESS BAR? fields = row_field_list GREATER tags = name_tag_list RBRACKET
+        { Ptyp_variant(fields, Closed, Some tags) }
+    | ext = extension
+        { Ptyp_extension ext }
+  )
+  { $1 }
+;
+
+object_type:
+  | mktyp(
+      LESS meth_list = meth_list GREATER
+        { let (f, c) = meth_list in Ptyp_object (f, c) }
+    | LESS GREATER
+        { Ptyp_object ([], Closed) }
+  )
+  { $1 }
+;
+
+delimited_type:
+  | object_type
+  | delimited_type_without_object_type
+    { $1 }
+;
+
 atomic_type:
-  | LPAREN core_type RPAREN
-      { $2 }
-  | LPAREN MODULE ext_attributes package_type RPAREN
-      { wrap_typ_attrs ~loc:$sloc (reloc_typ ~loc:$sloc $4) $3 }
+  | type_ = delimited_type 
+      { type_ }
+  | tys = actual_type_parameters
+    tid = mkrhs(type_longident)
+      { mktyp ~loc:$sloc (Ptyp_constr (tid, tys)) }
+  | tys = actual_type_parameters
+    HASH
+    cid = mkrhs(clty_longident)
+      { mktyp ~loc:$sloc (Ptyp_class (cid, tys)) }
```

Comment:
Why are the two cases above moved out of the common `mktyp( ... )` wrapper below? (My guess would be that they were different when you wanted to allow local-open on parameter tuples, but that you reverted this aspect and could in fact make the diff smaller and the code slightly more factorized by putting them back in the wrapper.)